x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/cluster/routing/allocation/DataTierAllocationDecider.java

```diff
@@ -65,8 +67,24 @@
         VALIDATOR, Setting.Property.Dynamic, Setting.Property.IndexScope);
     public static final Setting<String> INDEX_ROUTING_EXCLUDE_SETTING = Setting.simpleString(INDEX_ROUTING_EXCLUDE,
         VALIDATOR, Setting.Property.Dynamic, Setting.Property.IndexScope);
-    public static final Setting<String> INDEX_ROUTING_PREFER_SETTING = Setting.simpleString(INDEX_ROUTING_PREFER,
-        VALIDATOR, Setting.Property.Dynamic, Setting.Property.IndexScope);
+    public static final Setting<String> INDEX_ROUTING_PREFER_SETTING = new Setting<String>(new Setting.SimpleKey(INDEX_ROUTING_PREFER),
+        DataTierValidator::getDefaultTierPreference, Function.identity(), VALIDATOR, Setting.Property.Dynamic, Setting.Property.IndexScope) {
+        @Override
+        public String get(Settings settings) {
+            final Version idxVersion = IndexMetadata.SETTING_INDEX_VERSION_CREATED.get(settings);
+            if (idxVersion.onOrAfter(Version.V_7_13_0)) {
+                // 7.13.0+ indices can use the original setting value
+                return super.get(settings);
+            } else if (SearchableSnapshotsConstants.isPartialSearchableSnapshotIndex(settings)) {
+                // Partial searchable snapshot indices should be restricted to
+                // only data_frozen when reading the setting, or else validation fails.
+                return DATA_FROZEN;
+            } else {
+                // Otherwise pass through to the regular setting retrieval
+                return super.get(settings);
+            }
+        }
```

Comment:
@henningandersen I ended up having to change this for the backport. I'm not sure yet how I feel about it. Otherwise, a shared_cache searchable snapshot index created in 7.12.1 in a mixed-version cluster (where some clusters are 7.13.0) cause validation failures for the index, since we set it to `data_frozen,data_cold,data_warm,data_hot` in 7.12.1 and reject that in 7.13.0

Does this seem like a reasonable compromise for that case?

Reply:
@dakrone I do not really have a better workaround except to allow an explicit upgrade which does seems a bit excessive. I could be mildly worried about `Settings.Builder.put(Settings)` copying the underlying map rather than the setting values, meaning a copy would get the full string, but I have a hard time figuring out a scenario where this could cause issues. I would possibly prefer to avoid the version check in 7.x and just always return `DATA_FROZEN` for frozen indices.

Maybe @rjernst has input to this?

Reply:
I think the current code in the PR is the right approach. I don't think copying via settings builder will be an issue, whatever is in the map will be the same in the new map, there won't be any translation (we don't call get to copy the values, since we don't have the appropriate Setting objects).

Reply:
Okay, thanks for taking a look Ryan.

@henningandersen I pushed a change to make it unconditional instead of version-checked.